### PR TITLE
[Kotlin Migration] BackupManager

### PR DESCRIPTION
**Pull request by Akshay0701**

### CODE REVIEW

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
```
> Task :AnkiDroid:compilePlayReleaseJavaWithJavac
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:1039: error: no suitable method found for performBackupInBackground(String,Time)
        BackupManager.performBackupInBackground(getCol().getPath(), getCol().getTime());
                     ^
    method BackupManager.performBackupInBackground(String,int,boolean,Time) is not applicable
      (actual and formal argument lists differ in length)
    method BackupManager.performBackupInBackground(String,boolean,Time) is not applicable
      (actual and formal argument lists differ in length)
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
1 error
```

File: AnkiDroid/src/main/java/com/ichi2/anki/BackupManager.kt [hunk 1]
```diff
@@ -14,415 +14,397 @@
...
+import java.text.SimpleDateFormat
+import java.util.*
+import java.util.Collections.sort
+import java.util.zip.ZipEntry
+import java.util.zip.ZipOutputStream
+import kotlin.collections.ArrayList
+
+open class BackupManager {
+    @Throws(OutOfSpaceException::class)
+    fun performDowngradeBackupInForeground(path: String?): Boolean {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/BackupManager.kt:43]: non-null

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
```
com.ichi2.anki.BackupManagerTest > noBackupPerformedIfBackupAlreadyExists FAILED
    java.lang.NullPointerException
        at com.ichi2.anki.BackupManager$Companion.getBackups(BackupManager.kt:364)
        at com.ichi2.anki.BackupManager.performBackupInBackground(BackupManager.kt:70)
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:710)
        at java.base/java.lang.Thread.run(Thread.java:829)

com.ichi2.anki.BackupManagerTest > noBackupPerformedIfNoBackupNecessary FAILED
    java.lang.NullPointerException
        at com.ichi2.anki.BackupManager$Companion.getBackups(BackupManager.kt:364)
        at com.ichi2.anki.BackupManager.performBackupInBackground(BackupManager.kt:70)
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:710)
        at java.base/java.lang.Thread.run(Thread.java:829)

com.ichi2.anki.BackupManagerTest > testPassingSpy FAILED
    java.lang.NullPointerException
        at com.ichi2.anki.BackupManager$Companion.getBackups(BackupManager.kt:364)
        at com.ichi2.anki.BackupManager.performBackupInBackground(BackupManager.kt:70)
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:710)
        at java.base/java.lang.Thread.run(Thread.java:829)

com.ichi2.anki.BackupManagerTest > noBackupPerformedIfCollectionTooSmall FAILED
    java.lang.NullPointerException
        at com.ichi2.anki.BackupManager$Companion.getBackups(BackupManager.kt:364)
        at com.ichi2.anki.BackupManager.performBackupInBackground(BackupManager.kt:70)
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:710)
        at java.base/java.lang.Thread.run(Thread.java:829)

com.ichi2.anki.BackupManagerTest > passesIfNoBackupsAllowedAndForceIsEnabled FAILED
    java.lang.NullPointerException
        at com.ichi2.anki.BackupManager$Companion.getBackups(BackupManager.kt:364)
        at com.ichi2.anki.BackupManager.performBackupInBackground(BackupManager.kt:70)
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:710)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

**Comment by Akshay0701** [comment]
Yes, I am solving this error now, 
but confused on how to solve this..

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
Hmm... problem in testing,

we pass in "" as the path, the parent is null

But `File(null, "backups")` inside `getBackupDirectory` returns `C:\GitHub\Anki-Android-Git\Anki-Android\AnkiDroid\backup` so things work. We mostly use mocks here, so things pass

To fixup: 

And add a GitHub issue that the tests should be changed to use the temp directory, then fix the nullability. `.parent` shouldn't return null.

```
        @JvmStatic
        fun getBackups(colFile: File): Array<File?> {
            var files = getBackupDirectory(colFile.parentFile).listFiles()
            if (files == null) {
                files = arrayOfNulls(0)
            }
            val deckBackups = ArrayList<File?>(files.size)
            for (aktFile in files) {
                if (aktFile!!.name.replace("^(.*)-\\d{4}-\\d{2}-\\d{2}-\\d{2}-\\d{2}.(apkg|colpkg)$".toRegex(), "$1")
                    == colFile.name.replace(".anki2", "")
                ) {
                    deckBackups.add(aktFile)
                }
            }
            sort(deckBackups)
            val fileList = arrayOfNulls<File>(deckBackups.size)
            deckBackups.toArray(fileList)
            return fileList
        }
```

for now, accept null in `getBackupDirectory()`

**Pull request closed without merging**